### PR TITLE
Implement TensorExpression Product

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

This PR implements inner and outer products for working with `TensorExpression`s and adds accompanying unit tests. See **Further comments** regarding the tests.

The new `OuterProduct` struct defines an outer product of two expressions. If two operands of the `*` operator have a common index that needs to be contracted, the `OuterProduct` expression will be nested inside a `TensorContract` expression to take the inner product of the outer product.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

The unit tests are quite verbose, likely not very maintainable, and may be exhaustive in places they don't need to be. Please let me know if you would like me to gut or simplify any of the test cases.

The product test cases  are listed in the descriptions of the tester functions. Generally, they include tests for computing (1) inner products for all indices (e.g. `R_{ab} * S^{AB}`), (2) outer products for all indices (e.g. `R_{ab} * S_{cd}`), and (3) products resulting from a combination of inner and outer products of indices (e.g. `R_{ab} * S^{BC}`).

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]

File: src/DataStructures/Tensor/Expressions/Product.hpp
Comment (fmahebert) [inline src/DataStructures/Tensor/Expressions/Product.hpp]: For my understanding: is there a reason to favor writing `unsigned long` in the static_cast over `size_t`?

File: src/DataStructures/Tensor/Expressions/Product.hpp
Comment (fmahebert) [inline src/DataStructures/Tensor/Expressions/Product.hpp]: [nitpick, optional] This sentence "The component is computed by computing..." is a little awkward. I suggest rephrasing as "The function first computes the storage indices of the pair of components in the two RHS operand expressions, then multiplies the values at these storage indices to obtain the final component" ... or similar?

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (fmahebert) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: I suggest renaming `create_tensor` to something like `initialize_tensor` or `assign_values_to_tensor` (or any other suggestion you might have) to make it clearer that the tensor already exists and this function's job is putting numbers into the slots.

(also the create_tensor a few lines above for doubles)

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (fmahebert) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: I think this is slightly clearer phrased as "Replace tensor's value from `used_for_size` to a proper double" or similar... for me it took reading this twice to understand that by "instead of X" you meant "it is currently X, but I want it to be Y"

(And perhaps below "Replace tensor values so each `DataVector` element has a unique value" ... but if you rename `create_tensor` you might not need a comment there at all. Up to you!)

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (fmahebert) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: The index up/down-ness doesn't match on LHS and RHS here

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (fmahebert) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: Same comment about up/down...

Based on the code below, it looks like the RHS is correct? So I think the comment should change but everything else is correct?

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (fmahebert) [inline tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: Same comment about comment

**Comment by macedo22** [comment]
@fmahebert I believe I've addressed all your comments! If https://github.com/sxs-collaboration/spectre/pull/2771 is merged before this is, then I will update this PR by rebasing, overriding the virtual destructor, and simplifying the `operator*`.

File: tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp
Comment (macedo22) [reply tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp]: Good catch!

**Review by nilsdeppe — COMMENTED** [review]
I'm fine with you squashing once @fmahebert has had a chance to look at the fixups

**Comment by fmahebert** [comment]
Fixup looks good — please squash (now, or after waiting for the other PR to be merged, as you prefer)

**Comment by macedo22** [comment]
Once the other PR is merged, I planned on pushing a fixup commit for overriding the virtual destructor in `OuterProduct` and simplifying `operator*` to squash after it's reviewed, but if you would prefer the next time I push to just have this change already squashed because it's a small change, let me know and I'll do that instead

**Comment by macedo22** [comment]
I added 4b9bb63 for transparency, but can squash and push again once given the OK

**Comment by fmahebert** [comment]
OK with me 👍

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Comment by macedo22** [comment]
I'm closing this one with a :tada: because I'm really excited to see the last of the three operations working. There's still work to do, but I'm looking forward to the more complex expressions we'll now be able to write!

**Pull request merged**
